epinorm/norm.py

```diff
@@ -320,27 +474,170 @@ def _add_missing_columns(self):
+            country_ids[i] = country.get("id")
+
+            areas = location["areas"]
+            if len(areas) == 0:
+                continue
+
+            # try to match tokens in areas with the administrative_units file
+            admin_levels_found = []
+            admin_levels_country = []
+            if country.get("name") in admin_levels_table:
+                admin_levels_country = admin_levels_table[country.get("name")]
+
+                for admin_level in admin_levels_country:
+                    if admin_level["name"].lower() in areas:
+                        areas.remove(admin_level["name"].lower()) # this removes one occurance on purpose
+                        admin_levels_found.append(admin_level)
+            
+            # we couldn't identify any token, we must use the api to find all fields
+            if len(admin_levels_found) == 0:
+                result = self._find_full_locality(areas, country.get("name"))
+                localities[i] = result["locality"]
+                locality_osm_ids[i] = result["locality_osm_id"]
+                admin_level_1s[i] = result["admin_level_1"]
+                admin_level_1_ids[i] = result["admin_level_1_osm_id"]
+                continue
+            
+            # now we check if we already found the highest admin level (hence lowest value) of the location
+            min_admin_level_country = min(map(lambda x: x["admin_level"], admin_levels_country))
+            if country.get("name") in MIN_ADMIN_EXCEPTIONS:
+                min_admin_level_country = MIN_ADMIN_EXCEPTIONS[country.get("name")] 
+            my_min_admin_level = min(admin_levels_found, key=lambda x: x["admin_level"])
+            
+            if min_admin_level_country == my_min_admin_level["admin_level"]:
+                admin_level_1s[i] = my_min_admin_level["name"]
+                admin_level_1_ids[i] = my_min_admin_level["osm_id"]
+                admin_levels_found.remove(my_min_admin_level)
+
+                # now check if you have all the rest of the info you need
+                if len(areas) == 0 and len(admin_levels_found) > 0:
+                    localities[i] = admin_levels_found[0]["name"]
+                    locality_osm_ids[i] = admin_levels_found[0]["osm_id"]
+                    continue
+
+            # we got no more information
+            if len(areas) == 0 and len(admin_levels_found) == 0:
+                continue
+
+            # we are here if we haven't identified the highest admin boundary, or we might still have
+            # information on the city that wasn't matched with admin_boundaries
+
+            sorted(admin_levels_found, key=lambda x: x["admin_level"], reverse=True)
+            adminLevelsMatched = list(map(lambda x: x["name"], admin_levels_found))
+            admin_level = "" if admin_level_1s[i] is None else admin_level_1s[i]
+            second = ", ".join(adminLevelsMatched) + ", " + admin_level + ", " + country.get("name")
```

Comment:
Use snake_case for variable names.

```suggestion
            matched_admin_levels = list(map(lambda x: x["name"], admin_levels_found))
            admin_level = "" if admin_level_1s[i] is None else admin_level_1s[i]
            second = ", ".join(matched_admin_levels) + ", " + admin_level + ", " + country.get("name")
```